budgetportal/tests/test_prov_infra_projects.py

```diff
@@ -1,247 +1,959 @@
+        )
+        # Add ten projects
+        provinces = ["Eastern Cape", "Free State"]
+        for i in range(10):
+            if i < 5:
+                province = provinces[0]
             else:
-                status = self.statuses[1]
-                province = self.provinces[1]
-                source = self.sources[1]
-
-            ProvInfraProject.objects.create(
-                financial_year=self.fin_year,
-                IRM_project_id=i,
+                province = provinces[1]
+            project = ProvInfraProject.objects.create(IRM_project_id=i)
+            ProvInfraProjectSnapshot.objects.create(
+                irm_snapshot=self.irm_snapshot,
+                project=project,
                 name="Project {}".format(i),
-                department="Department {}".format(i),
-                local_municipality="Local {}".format(i),
-                district_municipality="District {}".format(i),
                 province=province,
-                status=status,
-                primary_funding_source=source,
-                main_contractor="Contractor {}".format(i),
-                principle_agent="Principle Agent {}".format(i),
-                program_implementing_agent="Program Agent {}".format(i),
-                other_parties="Service Provider: XXXX{}".format(i),
+                estimated_completion_date=date(year=2020, month=1, day=1)
             )
+        ProvInfraProjectIndex().reindex()
 
-    def test_projects_per_page(self):
-        response = self.client.get(self.url, format="json")
-        self.assertEqual(response.status_code, status.HTTP_200_OK)
+    def tearDown(self):
+        ProvInfraProjectIndex().clear()
+
+    def test_correct_numbers_showed(self):
+        selenium = self.selenium
+        selenium.get("%s%s" % (self.live_server_url, self.url))
+        num_of_projects = selenium.find_element_by_xpath(
+            '//*[@id="num-matching-projects-field"]'
+        ).text
+        num_of_projects = int(num_of_projects)
+        self.assertEqual(num_of_projects, 11)
+
+        num_of_items = len(
+            selenium.find_elements_by_xpath('//*[@id="result-list-container"]/a')
+        )
+        # Header is also counted
+        num_of_items = num_of_items - 1
+        self.assertEqual(num_of_projects, num_of_items)
+
+    def test_number_updated_after_search(self):
+        province = "Eastern Cape"
+        selenium = self.selenium
+        selenium.get("%s%s" % (self.live_server_url, self.url))
+        num_of_projects = selenium.find_element_by_xpath(
+            '//*[@id="num-matching-projects-field"]'
+        ).text
+        num_of_projects = int(num_of_projects)
+        self.assertEqual(num_of_projects, 11)
+
+        search_field = selenium.find_element_by_name("Infrastructure-Search")
+        search_button = selenium.find_element_by_xpath('//*[@id="Search-Button"]')
+        search_field.send_keys(province)
+        search_button.click()
+        time.sleep(self.timeout)
```

Comment:
never use time.sleep in GUI tests - rather use `wait.until`  - in this case use `text_to_be_present_in_element` as the condition https://selenium-python.readthedocs.io/waits.html

The best way to write these tests is to do
`wait.until(EC.text_to_be_present_in_element((By.CSS_SELECTOR, '#someid'), 12345))` for each check because then even if selenium is faster than our jquery, selenium will wait for each field (and it will normally just be super fast

Reply:
wait.until added.

https://github.com/vulekamali/datamanager/pull/213/commits/0809f4566bfa62d3e81e546bbd15d4096f5a42d5